zipkin/src/main/java/zipkin2/internal/ReadBuffer.java

```diff
@@ -0,0 +1,384 @@
+        | (readByteUnsafe() & 0xffL) << 48
+        | (readByteUnsafe() & 0xffL) << 56;
+    }
+
+    @Override public int pos() {
+      return buf.position();
+    }
+
+    @Override public int read(byte[] dst, int offset, int length) {
+      int toRead = checkReadArguments(dst, offset, length);
+      if (toRead == 0) return 0;
+      buf.get(dst, offset, toRead);
+      return toRead;
+    }
+
+    @Override public long skip(long maxCount) {
+      int skipped = Math.max(available(), (int) maxCount);
+      buf.position(buf.position() + skipped);
+      return skipped;
+    }
+
+    @Override public int available() {
+      return buf.remaining();
+    }
+  }
+
+  static final class Array extends ReadBuffer {
+    final byte[] buf;
+    int offset, length;
+
+    Array(byte[] buf, int offset, int length) {
+      this.buf = buf;
+      this.offset = offset;
+      this.length = length;
+    }
+
+    @Override final byte readByteUnsafe() {
+      return buf[offset++];
+    }
+
+    @Override final byte[] readBytes(int length) {
+      require(length);
+      byte[] result = new byte[length];
+      System.arraycopy(buf, offset, result, 0, length);
+      offset += length;
+      return result;
+    }
+
+    @Override public int read(byte[] dst, int offset, int length) {
+      int toRead = checkReadArguments(dst, offset, length);
+      if (toRead == 0) return 0;
+      System.arraycopy(buf, this.offset, dst, 0, toRead);
+      this.offset += toRead;
+      return toRead;
+    }
+
+    @Override boolean tryReadAscii(char[] destination, int length) {
+      for (int i = 0; i < length; i++) {
+        byte b = buf[offset + i];
+        if ((b & 0x80) != 0) return false;  // Not 7-bit ASCII character
+        destination[i] = (char) b;
+      }
+      offset += length;
+      return true;
+    }
+
+    @Override final String doReadUtf8(int length) {
+      String result = new String(buf, offset, length, UTF_8);
+      offset += length;
+      return result;
+    }
+
+    @Override short readShort() {
+      require(2);
+      return (short) ((buf[offset++] & 0xff) << 8 | (buf[offset++] & 0xff));
+    }
+
+    @Override int readInt() {
+      require(4);
+      int pos = this.offset;
+      this.offset = pos + 4;
+      return (buf[pos] & 0xff) << 24
+        | (buf[pos + 1] & 0xff) << 16
+        | (buf[pos + 2] & 0xff) << 8
+        | (buf[pos + 3] & 0xff);
+    }
+
+    @Override long readLong() {
+      require(8);
+      int pos = this.offset;
+      this.offset = pos + 8;
+      return (buf[pos] & 0xffL) << 56
+        | (buf[pos + 1] & 0xffL) << 48
+        | (buf[pos + 2] & 0xffL) << 40
+        | (buf[pos + 3] & 0xffL) << 32
+        | (buf[pos + 4] & 0xffL) << 24
+        | (buf[pos + 5] & 0xffL) << 16
+        | (buf[pos + 6] & 0xffL) << 8
+        | (buf[pos + 7] & 0xffL);
+    }
+
+    @Override long readLongLe() {
+      require(8);
+      int pos = this.offset;
+      this.offset = pos + 8;
+      return (buf[pos] & 0xffL)
```

Comment:
It's hardly better, but could create a method that takes 8 `byte` parameters to share the logic between buf / array

Reply:
didn't think about that either.. glad you are on the case!

Reply:
I think I saw in gson code that using a local ref to an array is faster than field. I'll bench first about using that approach.

Reply:
I was wrong about local array at least in JDK8. If any of the other approaches are more performant probably need to use a different JDK to notice. closing this topic for now.. thanks!

Reply:
I think the local variable helps with a volatile field, but probably nothing much for non-volatile. Anyways, `reverseBytes` is cool, now we don't have the duplication so this looks good.